### PR TITLE
Update DocumentOps docs

### PR DESCRIPTION
**Pull request by giarc3**
Trying to make ironoxide more easily understood from the docs alone. This PR focuses on the `document` functions and things related to them. 

I added examples for every `DocumentOps` function, which are compiled (but not run, because of async issues) when running `cargo test`, helping prove that they don't become out-of-date.

I need feedback on how the landing page of the docs site should flow. My current thought is that each section (Users, Groups, Documents), as well as a higher level (like Initialization) will provide a bare-bones example of the most important functions to get you up and running. A reader could start at the top and see how to initialize, create a user, create a group, and encrypt/decrypt a document right from that page. There will also be links to important functions for each section.

### CODE REVIEW

File: src/document/mod.rs [hunk 1]
```diff
@@ -46,14 +40,46 @@ impl ExplicitGrant {
     }
 }
 
+/// Parameters that can be provided when encrypting a new document.
+///
+/// If no document ID is provided, one will be generated for it. If no document name is provided, the document's
+/// name will be left empty.
```
Comment (ernieturner) [inline src/document/mod.rs:46]:
```suggestion
/// name will be left empty. Document IDs must be unique per segment.
```

File: src/document/mod.rs [hunk 2]
```diff
@@ -46,14 +40,46 @@ impl ExplicitGrant {
...
+    /// Constructs a new `DocumentEncryptOpts`.
+    ///
+    /// This requires either an `ExplicitGrant`, a `PolicyGrant`, or both. If only using one type
+    /// of grant, consider using [with_explicit_grants()](./struct.DocumentEncryptOpts.html#method.with_explicit_grants)
+    /// or [with_policy_grants()](./struct.DocumentEncryptOpts.html#method.with_policy_grants) instead.
+    ///
+    /// # Arguments
+    /// - `id` - Unique ID to use for the document. Note: this ID will **not** be encrypted.
+    /// - `name` - Non-unique name to use for the document. Note: this name will **not** be encrypted.
+    /// - `grants` - Grants that control who will be granted access to this encrypted document.
```
Comment (ernieturner) [inline src/document/mod.rs:67]:
```suggestion
    /// - `grants` - Grants that control who will be granted access to decrypt this document.
```

File: src/document/mod.rs [hunk 3]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
 }
 
 impl Default for DocumentEncryptOpts {
-    /// default to only sharing with the creator of the document
+    /// Constructs a `DocumentEncryptOpts` with common values.
+    ///
+    /// The document will have a generated ID and no name. Access to the document will be granted only to its creator.
```
Comment (ernieturner) [inline src/document/mod.rs:121]:
```suggestion
    /// The document will have a generated ID and no name. Access to decrypt the document will be granted only to its author.
```

File: src/document/mod.rs [hunk 4]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
         DocumentEncryptOpts::with_explicit_grants(None, None, true, vec![])
     }
 }
 
 #[async_trait]
 pub trait DocumentOps {
-    /// List all of the documents that the current user is able to decrypt.
+    /// Encrypts the provided document bytes.
+    ///
+    /// This will create an entry for the document in the IronCore service. Returns a `DocumentEncryptResult`
```
Comment (ernieturner) [inline src/document/mod.rs:131]: > This will create an entry for the document in the IronCore service.

Reading this, I'm not entirely sure what it means or what we want to convey here. Maybe we should just remove it entirely? I can't think of a great way to reword it to make it more obvious what we mean. I mostly just don't want people to think that we're storing the encrypted document or anything.

File: src/document/mod.rs [hunk 5]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
+    /// The document will have a generated ID and no name. Access to the document will be granted only to its creator.
     fn default() -> Self {
         DocumentEncryptOpts::with_explicit_grants(None, None, true, vec![])
     }
 }
 
 #[async_trait]
 pub trait DocumentOps {
-    /// List all of the documents that the current user is able to decrypt.
+    /// Encrypts the provided document bytes.
```
Comment (ernieturner) [inline src/document/mod.rs:129]:
```suggestion
    /// Encrypts the provided document bytes and optionally grants access to other users or groups to be able to decrypt the document.
```

File: src/document/mod.rs [hunk 6]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
+    /// # Examples
+    /// ```
+    /// # async fn run() -> Result<(), ironoxide::IronOxideErr> {
+    /// # use ironoxide::prelude::*;
+    /// # let sdk: IronOxide = unimplemented!();
+    /// use ironoxide::document::DocumentEncryptOpts;
+    /// let data = "secret data".as_bytes();
+    /// let encrypted = sdk.document_encrypt(data, &DocumentEncryptOpts::default()).await?;
+    /// # Ok(())
+    /// # }
```
Comment (ernieturner) [inline src/document/mod.rs:148]: Just curious, what do the `#` prefixes mean vs the lines that don't have them?

File: src/document/mod.rs [hunk 7]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
+    /// # let encrypted_data: Vec<u8> = vec![];
+    /// let decrypted_document = sdk.document_decrypt(&encrypted_data).await?;
+    /// let decrypted_data = decrypted_document.decrypted_data();
+    /// # Ok(())
+    /// # }
+    /// ```
+    async fn document_decrypt(&self, encrypted_document: &[u8]) -> Result<DocumentDecryptResult>;
+
+    /// Lists all of the IronCore encrypted documents that the calling user has been
+    /// granted access to.
```
Comment (ernieturner) [inline src/document/mod.rs:181]: Maybe we just remove the "IronCore" part here. I know that your goal here was that the unmanaged documents won't be returned here, but I'm not sure if that's worth specifically calling out? I dunno, I could be swayed either way, I was just trying to make it less vague as to what an "IronCore document" meant.
```suggestion
    /// Lists all of the encrypted documents that the calling user can decrypt
```

File: src/document/mod.rs [hunk 8]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
+    /// # use ironoxide::document::DocumentListMeta;
+    /// # let sdk: IronOxide = unimplemented!();
+    /// let document_data = sdk.document_list().await?;
+    /// let documents: Vec<DocumentListMeta> = document_data.result().to_vec();
+    /// # Ok(())
+    /// # }
     async fn document_list(&self) -> Result<DocumentListResult>;
 
-    /// Get the metadata for a specific document given its ID.
+    /// Returns the metadata for an IronCore encrypted document.
```
Comment (ernieturner) [inline src/document/mod.rs:195]:
```suggestion
    /// Returns the metadata for an encrypted document. Does not return the encrypted document bytes.
```

File: src/document/mod.rs [hunk 9]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
     ///
-    /// # Returns
-    /// `DocumentMetadataResult` with details about the requested document.
+    /// # Examples
+    /// ```
+    /// # async fn run() -> Result<(), ironoxide::IronOxideErr> {
+    /// # use ironoxide::prelude::*;
+    /// # let sdk: IronOxide = unimplemented!();
+    /// # let document_id: DocumentId = unimplemented!();
+    /// let document_data = sdk.document_get_metadata(&document_id).await?;
```
Comment (ernieturner) [inline src/document/mod.rs:206]:
```suggestion
    /// let document_meta = sdk.document_get_metadata(&document_id).await?;
```

File: src/document/mod.rs [hunk 10]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
+    /// # Ok(())
+    /// # }
+    /// ```
     async fn document_get_metadata(&self, id: &DocumentId) -> Result<DocumentMetadataResult>;
 
-    /// Attempt to parse the document ID out of an encrypted document.
+    /// Returns the document ID from the bytes of an IronCore encrypted document.
```
Comment (ernieturner) [inline src/document/mod.rs:212]:
```suggestion
    /// Returns the document ID from the bytes of an encrypted document.
```

File: src/document/mod.rs [hunk 11]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
+    ///
+    /// This is the same ID returned by `DocumentEncryptResult.id()`.
```
Comment (ernieturner) [inline src/document/mod.rs:214]:
```suggestion
    /// This is the same ID returned by `DocumentEncryptResult.id()` or `DocumentMetadataResult.id()`.
```

File: src/document/mod.rs [hunk 12]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
     ///
     /// # Arguments
-    /// - `encrypted_document` - Encrypted document bytes
+    /// - `encrypted_document` - Bytes of the encrypted document
+    ///
+    /// # Errors
+    /// Fails if the provided encrypted bytes have no header.
```
Comment (ernieturner) [inline src/document/mod.rs:220]:
```suggestion
    /// Fails if the provided bytes are not an encrypted document or have no header.
```

File: src/document/mod.rs [hunk 13]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
+    /// # let sdk: IronOxide = unimplemented!();
+    /// # let bytes: Vec<u8> = vec![];
+    /// let document_id = sdk.document_get_id_from_bytes(&bytes)?;
+    /// # Ok(())
+    /// # }
+    /// ```
     fn document_get_id_from_bytes(&self, encrypted_document: &[u8]) -> Result<DocumentId>;
 
-    /// Encrypt the provided document bytes.
+    /// Updates the contents of an existing IronCore encrypted document.
```
Comment (ernieturner) [inline src/document/mod.rs:234]:
```suggestion
    /// Updates the contents of an existing encrypted document while preserving the documents metadata and user and group access control.
```

File: src/document/mod.rs [hunk 14]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
-    ///       `grants` - List of users/groups to grant access to this document once encrypted
-    async fn document_encrypt(
-        &self,
-        document_data: &[u8],
-        encrypt_opts: &DocumentEncryptOpts,
-    ) -> Result<DocumentEncryptResult>;
-
-    /// Update the encrypted content of an existing document. Persists any existing access to other users and groups.
+    /// The new contents will be encrypted, and which users and groups are granted access
+    /// will remain unchanged.
```
Comment (ernieturner) [inline src/document/mod.rs:237]: Removing as this has been combined with the above.
```suggestion
```

File: src/document/mod.rs [hunk 15]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
         &self,
         id: &DocumentId,
         new_document_data: &[u8],
     ) -> Result<DocumentEncryptResult>;
 
-    /// Decrypts the provided encrypted document and returns details about the document as well as its decrypted bytes.
-    ///
-    /// # Arguments
-    /// - `encrypted_document` - Bytes of encrypted document. Should be the same bytes returned from `document_encrypt`.
+    /// Updates a document's name to a new value.
```
Comment (ernieturner) [inline src/document/mod.rs:260]:
```suggestion
    /// Modify or remove a document's name.
```

File: src/document/mod.rs [hunk 16]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
+    /// # }
+    /// ```
     async fn document_update_name(
         &self,
         id: &DocumentId,
         name: Option<&DocumentName>,
     ) -> Result<DocumentMetadataResult>;
 
-    /// Grant access to a document. Recipients of document access can be either users or groups.
+    /// Grants access to a document for a list of users and groups.
```
Comment (ernieturner) [inline src/document/mod.rs:286]:
```suggestion
    /// Grants decryption access to a document for the provided users and/or groups.
```

File: src/document/mod.rs [hunk 17]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
     /// # Arguments
-    /// `document_id` - id of the document whose access is is being modified
-    /// `grant_list` - list of grants. Elements represent either a user or a group.
+    /// `document_id` - Unique ID of the document whose access is being modified.
+    /// `grant_list` - List of users and groups to grant access to.
     ///
-    /// # Returns
-    /// Outer result indicates that the request failed either on the client or that the server rejected
-    /// the whole request. If the outer result is `Ok` then each individual grant to a user/group
-    /// either succeeded or failed.
```
Comment (ernieturner) [inline src/document/mod.rs:177]: I think this content is still pretty useful. It's worth telling users that this operation won't fail if all of the provided users/groups fail.

File: src/document/mod.rs [hunk 18]
```diff
@@ -84,113 +116,225 @@ impl DocumentEncryptOpts {
...
+    /// # }
+    /// ```
     async fn document_grant_access(
         &self,
         document_id: &DocumentId,
         grant_list: &Vec<UserOrGroup>,
     ) -> Result<DocumentAccessResult>;
 
-    /// Revoke access from a document. Revocation of document access can be either users or groups.
+    /// Revokes access to a document for a list of users and groups.
```
Comment (ernieturner) [inline src/document/mod.rs:315]:
```suggestion
    /// Revokes decryption access to a document for the provided users and/or groups.
```

File: src/document/mod.rs (hunk 1 above)
Comment (giarc3) [reply src/document/mod.rs:46]: The uniqueness is called out in the functions to create a DocumentEncryptOpts, so I would lean toward not including it here. Do you think it's needed in both places?
Or I guess I could also remove it from the functions and _only_ have it here?

File: src/document/mod.rs (hunk 2 above)
Comment (giarc3) [reply src/document/mod.rs:67]: I worded it this way to try to imply that you're granted a little more than just the ability to decrypt (like seeing the ID, when it was last updated, who else has access, etc.). But you could probably argue that that extra information isn't worth implying. Thoughts?

File: src/document/mod.rs (hunk 6 above)
Comment (giarc3) [reply src/document/mod.rs:148]: The `#` hides the line from the example. So boilerplate code that will be included in all of them (like the `async fn run`) or things where we're faking the typechecker on (like the `unimplemented!()`) won't have to show up for the users.

File: src/document/mod.rs (hunk 7 above)
Comment (giarc3) [reply src/document/mod.rs:181]: Yeah I want a way to make that distinction, but I'm fine with taking it out for clarity.

This is one place where "that the calling user can decrypt" technically isn't correct though (which is why I wrote a test for `document_list` in the PR). If you're the author of a document, even if you make `grant_to_author` false, it will still show up in your list.

File: src/document/mod.rs (hunk 11 above)
Comment (giarc3) [reply src/document/mod.rs:214]: While the addition is true, I was only calling out that it's the same as `DocumentEncryptResult.id()` to show that someone who just encrypted a file shouldn't call this, since they already have access to that data. Maybe I need to explicitly call that out.

So I think I'd rather leave out the addition to keep it focused on the intent of it.

File: src/document/mod.rs (hunk 13 above)
Comment (giarc3) [reply src/document/mod.rs:234]: Rust documentation guidelines seem to want a brief summary of the function followed by additional notes/information on the following lines. So even though it would flow fine as a single sentence, I think it makes it too long for a brief summary.

File: src/document/mod.rs (hunk 14 above)
Comment (giarc3) [reply src/document/mod.rs:237]: See above

File: src/document/mod.rs (hunk 17 above)
Comment (giarc3) [reply src/document/mod.rs:177]: That's a good point. I think I'll make a `# Errors` section for this one to call that out

File: src/document/mod.rs (hunk 15 above)
Comment (giarc3) [reply src/document/mod.rs:260]: I like it, though I'll change the verb tenses to match the standard/consistency.

File: src/document/mod.rs (hunk 1 above)
Comment (ernieturner) [reply src/document/mod.rs:46]: I'm fine not including it here, but I'd update the line in the other places to say more than just "Unique ID". Having clarity of what we mean by "unique" would be good I think since it only has to be unique within the segment.

File: src/document/mod.rs (hunk 2 above)
Comment (ernieturner) [reply src/document/mod.rs:67]: I suppose we could change it to say "who will have access to read and decrypt this document" or something like that, maybe?

File: src/document/mod.rs (hunk 7 above)
Comment (ernieturner) [reply src/document/mod.rs:181]: Ah right, that weird corner case. I'd probably be ok with ignoring that here. Alternatively something like "Lists all of the encrypted documents that the caller user can decrypt or authored" I suppose. We can wait for someone else to weigh in too.

File: src/document/mod.rs (hunk 11 above)
Comment (ernieturner) [reply src/document/mod.rs:214]: Yep, that's fine.

File: src/document/mod.rs (hunk 13 above)
Comment (ernieturner) [reply src/document/mod.rs:234]: Ok, 👍

File: src/document/mod.rs (hunk 1 above)
Comment (giarc3) [reply src/document/mod.rs]: I decided to have it in both places, since I agree that it makes it more clear

File: src/document/mod.rs (hunk 2 above)
Comment (giarc3) [reply src/document/mod.rs]: :+1:

File: src/document/mod.rs (hunk 3 above)
Comment (giarc3) [reply src/document/mod.rs]: Reworded slightly, but took the suggestion.

I agree that "author" makes the most sense for describing who made the document, but we should keep in mind that `AssociationType` refers to it as `Owner`, which conveys something slightly different

File: src/document/mod.rs (hunk 4 above)
Comment (giarc3) [reply src/document/mod.rs]: The intent of this was to differentiate it from unmanaged encryption, but I agree that it doesn't convey that very well. I think it's important to call out that we're storing _some_ information about their document, but clarify that the data is not stored.

I'll try to work on the wording

File: src/document/mod.rs (hunk 7 above)
Comment (giarc3) [reply src/document/mod.rs]: I think I'll try "Lists all of the encrypted documents that the calling user can read or decrypt."

File: src/document/mod.rs (hunk 8 above)
Comment (giarc3) [reply src/document/mod.rs]: I think I'll go with
```
/// Returns the metadata for an encrypted document.
///
/// This will not return the encrypted document bytes, as they are not stored by IronCore.
```

File: src/document/mod.rs (hunk 16 above)
Comment (giarc3) [reply src/document/mod.rs]: Done

File: src/document/mod.rs (hunk 18 above)
Comment (giarc3) [reply src/document/mod.rs]: Done

File: src/document/mod.rs (hunk 17 above)
Comment (giarc3) [reply src/document/mod.rs:177]: Done, but I could use some feedback on it

File: src/document/mod.rs (hunk 4 above)
Comment (giarc3) [reply src/document/mod.rs]: Did it. Let me know what you think

File: src/document/mod.rs [hunk 19]
```diff
@@ -84,113 +116,241 @@ impl DocumentEncryptOpts {
...
-    /// `grant_list` - list of grants. Elements represent either a user or a group.
+    /// `document_id` - Unique ID of the document whose access is being modified.
+    /// `grant_list` - List of users and groups to grant access to.
     ///
-    /// # Returns
-    /// Outer result indicates that the request failed either on the client or that the server rejected
-    /// the whole request. If the outer result is `Ok` then each individual grant to a user/group
-    /// either succeeded or failed.
+    /// # Errors
+    /// Fails only if the calling user does not have access to the document. If the response is an `Ok`, then the resulting
```
Comment (ernieturner) [inline src/document/mod.rs:301]: Given that the request to ironcore-id can also fail, this is a little simplified. It's maybe not worthwhile to try and enumerate all of the reasons it can fail. What if we said something like

```
This operation supports partial success. If any of the provided grants fail the `DocumentAccessResult` will indicate which grants succeeded and which failed, and it will provide explanations for each failure.
```

File: src/document/mod.rs (hunk 3 above)
Comment (ernieturner) [reply src/document/mod.rs]: Now you made me sad that we aren't consistent with `author` and `owner` everywhere :(

File: src/document/mod.rs (hunk 19 above)
Comment (giarc3) [reply src/document/mod.rs]: I added the part about partial success, removed the over-simplification, and tweaked the second sentence

File: src/document/mod.rs [hunk 20]
```diff
@@ -84,113 +116,241 @@ impl DocumentEncryptOpts {
...
+    /// # let sdk: IronOxide = unimplemented!();
+    /// # let encrypted_data: Vec<u8> = vec![];
+    /// let decrypted_document = sdk.document_decrypt(&encrypted_data).await?;
+    /// let decrypted_data = decrypted_document.decrypted_data();
+    /// # Ok(())
+    /// # }
+    /// ```
+    async fn document_decrypt(&self, encrypted_document: &[u8]) -> Result<DocumentDecryptResult>;
+
+    /// Lists all of the encrypted documents that the calling user can read or decrypt.
```
Comment (clintfred) [inline src/document/mod.rs:185]:
```suggestion
    /// Lists metadata for all of the encrypted documents that the calling user can read/decrypt.
```

File: src/document/mod.rs [hunk 21]
```diff
@@ -84,113 +116,241 @@ impl DocumentEncryptOpts {
...
     ///
-    /// # Returns
-    /// `DocumentMetadataResult` with details about the requested document.
+    /// # Examples
+    /// ```
+    /// # async fn run() -> Result<(), ironoxide::IronOxideErr> {
+    /// # use ironoxide::prelude::*;
+    /// # let sdk: IronOxide = unimplemented!();
+    /// # let document_id: DocumentId = unimplemented!();
+    /// let document_meta = sdk.document_get_metadata(&document_id).await?;
```
Comment (clintfred) [inline src/document/mod.rs:212]: Not sure how much value this example has. Doesn't show me anything I can't get from looking at the signature.

File: src/document/mod.rs [hunk 22]
```diff
@@ -84,113 +116,241 @@ impl DocumentEncryptOpts {
...
 
-    /// Grant access to a document. Recipients of document access can be either users or groups.
+    /// Grants decryption access to a document for the provided users and/or groups.
+    ///
+    /// Returns lists of successful and failed grants.
     ///
     /// # Arguments
-    /// `document_id` - id of the document whose access is is being modified
-    /// `grant_list` - list of grants. Elements represent either a user or a group.
+    /// `document_id` - Unique ID of the document whose access is being modified.
```
Comment (clintfred) [inline src/document/mod.rs:297]: bullet these to get them to appear on separate lines in HTML

File: src/document/mod.rs [hunk 23]
```diff
@@ -84,113 +116,241 @@ impl DocumentEncryptOpts {
...
 
-    /// Revoke access from a document. Revocation of document access can be either users or groups.
+    /// Revokes decryption access to a document for the provided users and/or groups.
+    ///
+    /// Returns lists of successful and failed revocations.
     ///
     /// # Arguments
-    /// `document_id` - id of the document whose access is is being modified
-    /// `revoke_list` - List of revokes. Elements represent either a user or a group.
+    /// `document_id` - Unique ID of the document whose access is being modified.
```
Comment (clintfred) [inline src/document/mod.rs:331]: bullet here

File: src/document/mod.rs [hunk 24]
```diff
@@ -84,113 +116,241 @@ impl DocumentEncryptOpts {
...
+    /// ```
+    /// # async fn run() -> Result<(), ironoxide::IronOxideErr> {
+    /// # use ironoxide::prelude::*;
+    /// # let sdk: IronOxide = unimplemented!();
+    /// # let document_id: DocumentId = unimplemented!();
+    /// # let users: Vec<UserId> = vec![];
+    /// use std::convert::TryFrom;
+    /// use ironoxide::document::UserOrGroup;
+    /// // from a list of UserIds, `users`
+    /// let users_or_groups: Vec<UserOrGroup> = users.iter().map(|user| user.into()).collect();
```
Comment (clintfred) [inline src/document/mod.rs:315]: This is a great example as the [UserId] -> [UserOrGroup] conversion can be confusing. It also makes me want to make that easier, somehow...

File: src/document/mod.rs [hunk 25]
```diff
@@ -84,113 +116,241 @@ impl DocumentEncryptOpts {
...
+    /// # }
+    /// ```
     async fn document_update_name(
         &self,
         id: &DocumentId,
         name: Option<&DocumentName>,
     ) -> Result<DocumentMetadataResult>;
 
-    /// Grant access to a document. Recipients of document access can be either users or groups.
+    /// Grants decryption access to a document for the provided users and/or groups.
```
Comment (clintfred) [inline src/document/mod.rs:292]: I wonder if we should talk more about partial success failure in the main comment block. Both here and for revoke.

File: src/document/mod.rs [hunk 26]
```diff
@@ -200,28 +360,6 @@ pub trait DocumentOps {
 
 #[async_trait]
 impl DocumentOps for crate::IronOxide {
-    async fn document_list(&self) -> Result<DocumentListResult> {
```
Comment (clintfred) [inline src/document/mod.rs:203]: What are these changes? Reordering maybe?

File: src/document/mod.rs (hunk 24 above)
Comment (giarc3) [reply src/document/mod.rs:317]: I agree that it can be confusing. Even worse is if you have a [UserId] and a [GroupId] and want to combine them into a [UserOrGroup]. I think it requires two mutable vecs and multiple lines. 
I would also love to make that process easier.

File: src/document/mod.rs (hunk 26 above)
Comment (giarc3) [reply src/document/mod.rs:203]: Yes, re-ordering. The order we put in the file changes the order of the docs, so I tried to make a logical progress of functions that started with encrypt/decrypt

File: src/document/mod.rs [hunk 27]
```diff
@@ -46,14 +40,46 @@ impl ExplicitGrant {
...
+    id: Option<DocumentId>,
+    name: Option<DocumentName>,
+    // at least one user/group must be included either explicitly or via a policy
+    grants: EitherOrBoth<ExplicitGrant, PolicyGrant>,
+}
+
 impl DocumentEncryptOpts {
+    /// Constructs a new `DocumentEncryptOpts`.
+    ///
+    /// This requires either an `ExplicitGrant`, a `PolicyGrant`, or both. If only using one type
```
[the remark this answers is not in the document]
Comment (giarc3) [reply src/document/mod.rs]: Done

File: src/document/mod.rs [hunk 28]
```diff
@@ -46,14 +40,46 @@ impl ExplicitGrant {
...
+    /// of grant, consider using [with_explicit_grants()](./struct.DocumentEncryptOpts.html#method.with_explicit_grants)
```
[the remark this answers is not in the document]
Comment (giarc3) [reply src/document/mod.rs]: Done

File: src/document/mod.rs [hunk 29]
```diff
@@ -46,14 +40,46 @@ impl ExplicitGrant {
...
+    /// or [with_policy_grants()](./struct.DocumentEncryptOpts.html#method.with_policy_grants) instead.
```
[the remark this answers is not in the document]
Comment (giarc3) [reply src/document/mod.rs]: Done

File: src/document/mod.rs [hunk 30]
```diff
@@ -70,6 +96,12 @@ impl DocumentEncryptOpts {
         }
     }
 
+    /// Constructs a new `DocumentEncryptOpts` with access granted by a policy.
+    ///
+    /// # Arguments
+    /// - `id` - Unique ID to use for the document. Note: this ID will **not** be encrypted.
+    /// - `name` - Non-unique name to use for the document. Note: this name will **not** be encrypted.
+    /// - `grants` - Policy to determine which users and groups will have access to read and decrypt this document.
```
[the remark this answers is not in the document]
Comment (giarc3) [reply src/document/mod.rs]: Fixed the error and linked to the policy module

File: src/document/mod.rs [hunk 31]
```diff
@@ -84,113 +116,241 @@ impl DocumentEncryptOpts {
...
 
 #[async_trait]
 pub trait DocumentOps {
-    /// List all of the documents that the current user is able to decrypt.
+    /// Encrypts the provided document bytes.
     ///
-    /// # Returns
-    /// `DocumentListResult` struct with vec of metadata about each document the user can decrypt.
+    /// Returns a `DocumentEncryptResult` which contains document metadata as well as the `encrypted_data`,
+    /// which is the only thing that must be passed to [document_decrypt()](trait.DocumentOps.html#tymethod.document_decrypt)
```
[the remark this answers is not in the document]
Comment (giarc3) [reply src/document/mod.rs]: I've gone back on forth on my preference. I went ahead and removed the `()` from our links in every file

File: src/document/mod.rs [hunk 32]
```diff
@@ -84,113 +116,241 @@ impl DocumentEncryptOpts {
...
+    /// # Arguments
+    /// - `document_data` - Bytes of the document to encrypt
+    /// - `encrypt_opts` - Document encryption parameters. Default values are provided with `DocumentEncryptOpts::default()`.
+    ///
+    /// # Examples
+    /// ```
+    /// # async fn run() -> Result<(), ironoxide::IronOxideErr> {
+    /// # use ironoxide::prelude::*;
+    /// # let sdk: IronOxide = unimplemented!();
+    /// use ironoxide::document::DocumentEncryptOpts;
```
[the remark this answers is not in the document]
Comment (giarc3) [reply src/document/mod.rs]: Sure. Hidden

File: src/document/mod.rs (hunk 20 above)
Comment (giarc3) [reply src/document/mod.rs]: I added the "metadata for", but I prefer to keep the "or" over the slash

File: src/document/mod.rs (hunk 21 above)
Comment (giarc3) [reply src/document/mod.rs:214]: We discussed this elsewhere, but to recap, I'd prefer to leave this in as I think it gives an "alternate summary" of how to easily use the function
I also expanded this example slightly by showing the creation of the `DocumentId` from a string

File: src/document/mod.rs (hunk 25 above)
Comment (giarc3) [reply src/document/mod.rs:295]: I think that the "Returns lists of successful and failed grants" implies the partial success, and the `Errors` section clarifies it

File: src/document/mod.rs (hunk 22 above)
Comment (giarc3) [reply src/document/mod.rs]: Good catch. Fixed

File: src/document/mod.rs (hunk 23 above)
Comment (giarc3) [reply src/document/mod.rs]: Fixed
